File: 8-string-to-integer-atoi/string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        int index = 0, sign = 1, total = 0;

        // Step 1: Skip leading whitespaces
        while (index < s.length() && s.charAt(index) == ' ') {
            index++;
        }

        // Step 2: Check for '+' or '-'
        if (index < s.length() && (s.charAt(index) == '+' || s.charAt(index) == '-')) {
            sign = (s.charAt(index) == '-') ? -1 : 1;
            index++;
        }

        // Step 3: Read in digits until a non-digit character is encountered
        while (index < s.length() && Character.isDigit(s.charAt(index))) {
            int digit = s.charAt(index) - '0';

            // Step 4: Check for integer overflow
            if (total > Integer.MAX_VALUE / 10 || (total == Integer.MAX_VALUE / 10 && digit > Integer.MAX_VALUE % 10)) {
                return (sign == 1) ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }

            total = total * 10 + digit;
            index++;
        }

        return total * sign;
    }
}

File: 8-string-to-integer-atoi/string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(std::string s) {
        int i = 0;
        int sign = 1;
        long result = 0;

        // Step 1: Ignore leading whitespace
        while (i < s.size() && s[i] == ' ') {
            i++;
        }

        // Step 2: Check for sign
        if (i < s.size() && (s[i] == '-' || s[i] == '+')) {
            sign = (s[i++] == '-') ? -1 : 1;
        }

        // Step 3: Read digits until a non-digit character is encountered
        while (i < s.size() && isdigit(s[i])) {
            result = result * 10 + (s[i++] - '0');

            // Step 4: Check if result is within the 32-bit signed integer range
            if (result * sign > INT_MAX) {
                return INT_MAX;
            } else if (result * sign < INT_MIN) {
                return INT_MIN;
            }
        }

        return result * sign;
    }
};